api/model/src/main/java/org/projectnessie/model/Namespace.java

```diff
@@ -165,36 +168,24 @@ public static Namespace of(Map<String, String> properties, String... elements) {
       return EMPTY_NAMESPACE;
     }
 
-    for (String e : elements) {
-      if (e == null) {
-        throw new IllegalArgumentException(
-            String.format(
-                "Namespace '%s' must not contain a null element.", Arrays.toString(elements)));
-      }
-      if (e.isEmpty()) {
-        throw new IllegalArgumentException(
-            String.format(
-                "Namespace '%s' must not contain an empty element.", Arrays.toString(elements)));
-      }
-      if (e.chars().anyMatch(i -> i < FIRST_ALLOWED_KEY_CHAR)) {
-        throw new IllegalArgumentException(
-            String.format(
-                "Namespace '%s' must not contain characters less than 0x%2h.",
-                Arrays.toString(elements), FIRST_ALLOWED_KEY_CHAR));
-      }
-    }
-
-    if (DOT_STRING.equals(elements[elements.length - 1])) {
-      throw new IllegalArgumentException(
-          String.format(ERROR_MSG_TEMPLATE, Arrays.toString(elements)));
-    }
-
     return ImmutableNamespace.builder()
         .elements(Arrays.asList(elements))
         .properties(properties)
         .build();
   }
 
+  @Value.Check
+  protected void validate() {
+    Elements.super.validate("Namespace");
+
+    int elementCount = getElementCount();
+    List<String> elements = getElements();
+    if (elementCount > 0 && ".".equals(elements.get(elementCount - 1))) {
+      throw new IllegalArgumentException(
+          format("Namespace '%s' must not contain a '.' element", elements));
```

Comment:
One more nit, sorry: the constant `ERROR_MSG_TEMPLATE` is still in use, and it's used for a similar error message, see `Namespace#parse()`. I'd suggest to remove that usage and delete the constant (since the check will be done later on by  this method).

Reply:
done